### PR TITLE
[SPARK-6785][SQL] fix DateTimeUtils for dates before 1970

### PR DESCRIPTION
**Pull request by ckadner**
Hi Michael, 
this Pull-Request is a follow-up to [PR-6242](https://github.com/apache/spark/pull/6242). I removed the two obsolete test cases from the HiveQuerySuite and deleted the corresponding golden answer files.
Thanks for your review!

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

**Comment by ckadner** [comment]
@marmbrus - Hi Michael, please let me know when you have time to review my pull-request. This PR no longer merges, so I would close this and create a new PR on top of the latest code in apache/spark master (as opposed to pushing one merge and hundreds of unrelated changes to this PR)?

**Comment by squito** [comment]
Hi @ckadner.  I'm not the best person to review this, but just randomly saw your comment.  If you merge with master, and then push to your same branch, the diff here on github will still be fine.  It'll add one merge commit, but not all the other commits that got pulled.  Also github will still show the right diff, so reviewers will still be able to focus on the right bits.

**Comment by marmbrus** [comment]
+1 to @squito 's suggestion, and I'll add that you can also force push if you do a rebase instead of a merge.  Eitherway, github only is going to look at whatever the head commit of the branch you used to open this PR.  When we merge to spark master we have a script that squash commits, so it doesn't really matter if the tree is messy.  There will only be one clean commit no matter what the PR looks like.

I should be able to look at it this week (and I've marked myself as the shepherd on JIRA so it should show up on my dashboard when its ready).

**Comment by ckadner** [comment]
@squito - thx 4 your comments!  :+1: 

@marmbrus - back to you -- thx

**Comment by marmbrus** [comment]
Thanks!  Merging to master.

**Comment by ckadner** [comment]
Thanks Michael!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 1]
```diff
@@ -324,20 +324,6 @@ class HiveQuerySuite extends HiveComparisonTest with BeforeAndAfter {
       | FROM src LIMIT 1
     """.stripMargin)
 
-  createQueryTest("Date comparison test 2",
-    "SELECT CAST(CAST(0 AS timestamp) AS date) > CAST(0 AS timestamp) FROM src LIMIT 1")
```
Comment (adrian-wang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: @ckadner why do we remove this?
I think we have handled time zone issue in `beforeAll()`

Comment (ckadner) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: @adrian-wang at the time of [this fix](https://issues.apache.org/jira/browse/SPARK-6785), Spark was using Hive v0.13, which suffered from the same bug ([HIVE-10178](https://issues.apache.org/jira/browse/HIVE-10178)) and this test case did no longer test for equivalent behavior between Hive and Spark. Please see https://github.com/apache/spark/pull/6242#issuecomment-103670515 (and following) for the discussion about removing this test case.
Since Spark is now on Hive 1.2.1 and [HIVE-10178](https://issues.apache.org/jira/browse/HIVE-10178) was fixed in Hive 1.2, we could consider re-introducing this test case to the `HiveQuerySuite` (@marmbrus ?)

Comment (adrian-wang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: but these cases are non-negative?

Comment (adrian-wang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: I mean, the bug of Hive only affects those whose `daysSinceEpoch` is negative, but test cases here should yield Date values whose `daysSinceEpoch` is 0, so these cases shouldn't be affected by this bug.

Comment (ckadner) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: depends on the system time zone, it should be ...
- West of GMT `daysSinceEpoch` is `-1`
- East of GMT `daysSinceEpoch` is `0`
```
 "CAST(0 AS timestamp)" --> "1 Jan 1970 00:00:00 GMT"
 "CAST(0 AS timestamp)" --> "31 Dec 1969 16:00:00 PST"
```

Comment (adrian-wang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: That's true, but there is a setting in beforeAll, so spark should produce same result in regard of your system timezone.

Comment (ckadner) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]:
```
  override def beforeAll() {
    ...
    TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"))
```

- before my fix, both Spark (1.4) and Hive (0.13) returned `true`
- after my fix, Spark (1.5) returned `false` and Hive (0.13) returned `true` (that's why we removed it)
- since Spark (1.5) later moved to Hive 1.2.1 now, both Spark and Hive return `false`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 2]
```diff
@@ -48,4 +49,41 @@ class DateTimeUtilsSuite extends SparkFunSuite {
     val t2 = DateTimeUtils.toJavaTimestamp(DateTimeUtils.fromJulianDay(d1, ns1))
     assert(t.equals(t2))
   }
+
+  test("SPARK-6785: java date conversion before and after epoch") {
+    def checkFromToJavaDate(d1: Date): Unit = {
+      val d2 = DateTimeUtils.toJavaDate(DateTimeUtils.fromJavaDate(d1))
+      assert(d2.toString === d1.toString)
```
Comment (JoshRosen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: Shouldn't it also be the case that `d1 === d2`?

Comment (ckadner) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: No. The Date object is internally represented by a milliseconds since epoch -- which will not match for the use cases we are testing. We only want the year-month-day portion of the Date to match, which we get out via Date.toString()

Comment (JoshRosen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: Per the [Javadoc](https://docs.oracle.com/javase/6/docs/api/java/sql/Date.html) for `java.sql.Date`:

> To conform with the definition of SQL DATE, the millisecond values wrapped by a java.sql.Date instance must be 'normalized' by setting the hours, minutes, seconds, and milliseconds to zero in the particular time zone with which the instance is associated.

Given this, I'd assume that the value returned by `Date.valueOf("1970-01-01").getTime` is the canonical timestamp value for the date "1970-01-01" in the caller's JVM and that a roundtrip conversion would therefore return that exact timestamp.

For `new Date(df1.parse("1970-01-01 00:00:01").getTime)`, on the other hand, I think that the timestamp isn't normalized and thus I'm not sure that this is considered a "valid" `java.sql.Date` instance.

Basically I'm assuming that there'd only be one canonical `java.sql.Date` representation for a given date within a single JVM.

Comment (ckadner) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: Well, yes, but the code under test is the round-trip conversion: _Java_ Date -> _SQL_ Date -> _Java_ Date. 

Since "valid" ```java.util.Date``` instances do carry time information I included these non-normalized dates to test the boundaries of the conversion code in ```DateTimeUtils.scala```

To make the test case more concise, we could split out the _normalized_ date examples and test them with ```===```, but I would not want to eliminate the _non-normalized_ date(-time) tests.

Btw, the example in the issue description for [SPARK-6785](https://issues.apache.org/jira/browse/SPARK-6785) used a non-normalized Date as well.
```
scala> val d = new Date(100)
d: java.sql.Date = 1969-12-31

scala> DateUtils.toJavaDate(DateUtils.fromJavaDate(d))
res1: java.sql.Date = 1970-01-01
```
